### PR TITLE
Add course run endpoint

### PR DESCRIPTION
**Pull request by bderusha**
@clintonb @peter-fogg please review

### CODE REVIEW

File: course_discovery/apps/course_metadata/admin.py [hunk 1]
```diff
@@ -1,8 +1,8 @@
 from django.contrib import admin
 
 from course_discovery.apps.course_metadata.models import (
-    Seat, Image, Video, LevelType, Subject, Prerequisite, ExpectedLearningItem, Course, CourseRun, Organization, Person,
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/admin.py:4]: We have a 120 character limit. This file does not need to be changed.

File: course_discovery/apps/course_metadata/models.py [hunk 2]
```diff
@@ -58,18 +69,14 @@ class Prerequisite(AbstractNamedModel):
...
-    def __str__(self):
-        return self.value
+    pass
 
 
-class SyllabusItem(TimeStampedModel):
+class SyllabusItem(AbstractValueModel):
     """ SyllabusItem model. """
     parent = models.ForeignKey('self', blank=True, null=True, related_name='children')
-    value = models.CharField(max_length=255)
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py:72]: This value shouldn't necessarily be unique across all instances. For example, say I have the item, "Learn dynamic programming". This item, for my course, should be in two separate portions of my syllabus (e.g. two different parents). Making value unique breaks this.

File: course_discovery/apps/course_metadata/models.py [hunk 3]
```diff
@@ -23,6 +23,17 @@ class Meta(object):
         abstract = True
 
 
+class AbstractValueModel(TimeStampedModel):
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/models.py:26]: Make these model changes part of a separate commit.

File: course_discovery/apps/api/serializers.py [hunk 4]
```diff
@@ -86,6 +126,35 @@ class Meta(object):
...
+class CourseRunSerializer(TimestampModelSerializer):
+    content_language = serializers.SlugRelatedField(read_only=True, slug_field='code', source='language')
+    transcript_languages = serializers.SlugRelatedField(many=True, read_only=True, slug_field='code')
+    image = ImageSerializer()
+    video = VideoSerializer()
+    seats = SeatSerializer(many=True)
+    syllabus = serializers.SerializerMethodField()
+    instructors = PersonSerializer(many=True)
+    staff = PersonSerializer(many=True)
+    effort = serializers.SerializerMethodField()
```
Comment (clintonb) [inline course_discovery/apps/api/serializers.py]: Why bother with nesting this? Just leave it flat, as it already is on the model.

File: course_discovery/apps/api/serializers.py [hunk 5]
```diff
@@ -86,6 +126,35 @@ class Meta(object):
...
+
+    def get_effort(self, obj):
+        return EffortSerializer(obj).data
+
+    def get_syllabus(self, obj):
+        syllabus = obj.syllabus
+        if syllabus:
+            return SyllabusSerializer(syllabus.children, many=True).data
```
Comment (clintonb) [inline course_discovery/apps/api/serializers.py]: I don't see any recursion, so this won't be robust enough to completely replace a syllabus tree with more than one level.

File: course_discovery/apps/api/v1/views.py [hunk 6]
```diff
@@ -109,3 +111,21 @@ def list(self, request, *args, **kwargs):
     def retrieve(self, request, *args, **kwargs):
         """ Retrieve details for a course. """
         return super(CourseViewSet, self).retrieve(request, *args, **kwargs)
+
+
+class CourseRunViewSet(viewsets.ReadOnlyModelViewSet):
+    """ CourseRun resource. """
+    lookup_field = 'key'
+    lookup_value_regex = COURSE_RUN_ID_REGEX
+    queryset = CourseRun.objects.all().order_by(Lower('key'))
```
Comment (peter-fogg) [inline course_discovery/apps/api/v1/views.py:120]: Does this need to be ordered?

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.data, CourseRunSerializer(self.course_run).data)
+
+    def test_list(self):
+        """ Verify the endpoint returns a list of all catalogs. """
+        url = reverse('api:v1:course_run-list')
+
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 200)
+        self.assertListEqual(response.data['results'], CourseRunSerializer(CourseRun.objects.all(), many=True).data)
```
Comment (peter-fogg) [inline course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: As the converse of my other question, if we are guaranteeing the ordering of course runs in the response then we should test against the ordered queryset here.

File: course_discovery/apps/api/serializers.py (hunk 4 above)
Comment (bderusha) [inline course_discovery/apps/api/serializers.py]: Implementing the spec as agreed :)  If we want to change the API spec that's a different conversation.

Comment (clintonb) [inline course_discovery/apps/api/serializers.py]: Let's change the spec. It's our spec. We can change it. No sense in making the API more complicated than it needs to be just because we missed this in the review.

File: course_discovery/apps/api/tests/test_serializers.py [hunk 8]
```diff
@@ -136,3 +193,69 @@ def test_data(self):
...
+    def test_data(self):
+        course_run = CourseRunFactory()
+        serializer = EffortSerializer(course_run)
+
+        expected = {
+            'min': course_run.min_effort,
+            'max': course_run.max_effort
+        }
+
+        self.assertDictEqual(serializer.data, expected)
```
Comment (peter-fogg) [inline course_discovery/apps/api/tests/test_serializers.py]: Just as a minor nitpick, `assertEqual` will call the type-specific version when appropriate. Same for the `assertListEqual` call below and other calls to `assertDictEqual`.

**Comment by peter-fogg** [comment]
A couple small things other than what @clintonb noted; looks good otherwise.

File: course_discovery/apps/api/serializers.py (hunk 5 above)
Comment (bderusha) [inline course_discovery/apps/api/serializers.py]: Agreed.  While our data model allows for as much nesting as possible, our spec currently only expects the title and contents for now, which seems to handle most cases.  Since drf serializers do not handle recursion well I didn't want to over engineer for a case we do not need to handle presently.

File: course_discovery/apps/api/v1/views.py (hunk 6 above)
Comment (bderusha) [inline course_discovery/apps/api/v1/views.py:120]: A valid question.  It is a copy of the logic from the `CourseViewSet` and I honestly did not give the query much thought.  Thinking about it now, however, I believe that given the fact that the data is going to be ordered by the database regardless we might as well be in control rather than letting the database do whatever it wants (which can be inconsistent).

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py (hunk 7 above)
Comment (bderusha) [inline course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: Point taken.  Will update.

File: course_discovery/apps/api/serializers.py (hunk 5 above)
Comment (clintonb) [inline course_discovery/apps/api/serializers.py]: Actually, we kicked the syllabus out of the first release. This shouldn't be here.

File: course_discovery/apps/api/serializers.py [hunk 9]
```diff
@@ -86,6 +126,35 @@ class Meta(object):
         )
 
 
+class CourseRunSerializer(TimestampModelSerializer):
+    content_language = serializers.SlugRelatedField(read_only=True, slug_field='code', source='language')
```
Comment (clintonb) [inline course_discovery/apps/api/serializers.py:113]: Hindsight: `CourseRun.language` should be a many-to-many to account for the fact that course can be in multiple languages.

File: course_discovery/apps/api/tests/test_serializers.py (hunk 8 above)
Comment (clintonb) [inline course_discovery/apps/api/tests/test_serializers.py]: This is true, but the output is different. I find that the type-specific methods do a much better job of displaying diffs when there is a mismatch.

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+from course_discovery.apps.core.tests.factories import UserFactory, USER_PASSWORD
+from course_discovery.apps.course_metadata.tests.factories import CourseRunFactory
+from course_discovery.apps.course_metadata.models import CourseRun
+
+
+class CourseRunViewSetTests(APITestCase):
+    def setUp(self):
+        super(CourseRunViewSetTests, self).setUp()
+        self.user = UserFactory(is_staff=True, is_superuser=True)
+        self.client.login(username=self.user.username, password=USER_PASSWORD)
```
Comment (clintonb) [inline course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: FYI: Since you are using `APITestCase`, you can also use `self.client.force_authenticate(self.user)`. See http://www.django-rest-framework.org/api-guide/testing/#force_authenticateusernone-tokennone.

File: course_discovery/apps/api/tests/test_serializers.py (hunk 8 above)
Comment (peter-fogg) [inline course_discovery/apps/api/tests/test_serializers.py]: The type-specific methods are called: https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertEqual

Comment (clintonb) [inline course_discovery/apps/api/tests/test_serializers.py]: Great. That was finally added in Python 3.1...and Python 2.7. Weird.

**Comment by clintonb** [comment]
Looks good, aside from outstanding comments.

File: course_discovery/apps/course_metadata/models.py (hunk 2 above)
Comment (bderusha) [inline course_discovery/apps/course_metadata/models.py:72]: Fixed.

File: course_discovery/apps/course_metadata/models.py (hunk 3 above)
Comment (bderusha) [inline course_discovery/apps/course_metadata/models.py:26]: Fixed.

File: course_discovery/apps/api/serializers.py (hunk 9 above)
Comment (bderusha) [inline course_discovery/apps/api/serializers.py:113]: I believe we agreed that a course run is only being run in a single language.

Comment (clintonb) [inline course_discovery/apps/api/serializers.py:113]: Based on our upcoming partnerships, that can't/won't be true for long. We should plan ahead for course content to be translated.

Comment (bderusha) [inline course_discovery/apps/api/serializers.py:113]: Based on offline discussion we are okay with adding that functionality once we have supporting use cases

**Comment by bderusha** [comment]
Cool.  @peter-fogg @clintonb thumbs?

**Comment by peter-fogg** [comment]
:+1:

**Comment by clintonb** [comment]
:+1:
